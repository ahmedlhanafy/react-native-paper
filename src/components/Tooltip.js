/* @flow */

import * as React from 'react';
import { Dimensions, StyleSheet, UIManager, View } from 'react-native';
import * as Colors from '../styles/colors';
import { APPROX_STATUSBAR_HEIGHT } from '../constants';
import Text from './Typography/Text';
import Portal from './Portal/Portal';
import type { ViewStyleProp } from 'react-native/Libraries/StyleSheet/StyleSheet';

type Props = {|
  /**
   * Tooltip title.
   */
  title: string,
  /**
   * Tooltip reference node.
   */
  children: React.node,
  /**
   * Delay in ms before onLongPress is called.
   */
  delayLongPress?: number,
  /**
   * Style that is passed to the children wrapper.
   */
  style?: ViewStyleProp,
|};

type State = {
  childrenDimensions: {
    x: number,
    y: number,
    width: number,
    height: number,
  },
  tooltipDimensions: {
    width: number,
    height: number,
  },
  tooltipVisible: boolean,
  tooltipOpacity: number,
};

class Tooltip extends React.Component<Props, State> {
  static defaultProps = { delayLongPress: 500 };

  state = {
    childrenDimensions: {
      x: null,
      y: null,
      width: null,
      height: null,
    },
    tooltipDimensions: {
      width: null,
      height: null,
    },
    tooltipVisible: false,
    tooltipOpacity: 0,
  };

  componentWillUnmount() {
    this.clearTimeouts();
  }

  onChildrenLayout = ({ nativeEvent: { target } }) => {
    setTimeout(() => {
      UIManager.measure(target, (_x, _y, width, height, pageX, pageY) => {
        this.setState({
          childrenDimensions: {
            x: pageX,
            y: pageY,
            width,
            height,
          },
        });
      });
    }, 500);
  };

  onTooltipLayout = () => {
    this.tooltip.measureInWindow((_x, _y, width, height) => {
      const {
        tooltipDimensions: { width: tooltipWidth, height: tooltipHeight },
      } = this.state;

      // Prevent rerender after getting tooltip dimensions
      if (!tooltipWidth && !tooltipHeight)
        this.setState({
          tooltipDimensions: {
            width,
            height,
          },
          tooltipOpacity: 0.9,
        });
    });
  };

  showTooltip = () => this.setState({ tooltipVisible: true });

  hideTooltip = () => this.setState({ tooltipVisible: false });

  clearTimeouts = () => clearTimeout(this.longPressTimeout);

  onTouchStart = () => {
    const { delayLongPress } = this.props;

    this.longPressTimeout = setTimeout(() => {
      this.showTooltip();
    }, delayLongPress);
  };

  onTouchEndCapture = () => {
    const { delayLongPress } = this.props;
    const { tooltipVisible } = this.state;

    this.clearTimeouts();
    this.longPressTimeout = setTimeout(() => {
      tooltipVisible && this.hideTooltip();
    }, delayLongPress);
  };

  onTouchCancel = () => {
    this.clearTimeouts();
    this.hideTooltip();
  };

  getTooltipXPosition = () => {
    const {
      childrenDimensions: { x, width },
      tooltipDimensions: { width: tooltipWidth },
    } = this.state;

    const centerDistanceFromChildren = x + (width - tooltipWidth) / 2;
    const { width: layoutWidth } = Dimensions.get('window');

    // Does it overflow to the right? If so, subtracts tooltipWidth
    if (centerDistanceFromChildren + tooltipWidth > layoutWidth) {
      return x + width - tooltipWidth;
    }
    // Does it overflow to the left? If so, starts from children start position
    else if (centerDistanceFromChildren < 0) {
      return x;
    }

    return centerDistanceFromChildren;
  };

  getTooltipYPosition = () => {
    const {
      childrenDimensions: { y, height },
      tooltipDimensions: { height: tooltipHeight },
    } = this.state;

    const { height: layoutHeight } = Dimensions.get('window');

    // Does it overflow to the bottom? If so, subtracts the tooltip height,
    // the marginTop applied to the tooltip and the status bar height.
    if (y + height + tooltipHeight + APPROX_STATUSBAR_HEIGHT > layoutHeight) {
      return (
        y - tooltipHeight - styles.tooltip.marginTop - APPROX_STATUSBAR_HEIGHT
      );
    }

    return y + tooltipHeight;
  };

  render() {
    const { children, title, style, ...rest } = this.props;
    const childElement = React.Children.only(children);
    const { tooltipVisible, tooltipOpacity } = this.state;

    return (
      <View style={style}>
        <Portal>
          {tooltipVisible && (
            <View
              ref={el => {
                this.tooltip = el;
              }}
              style={[
                styles.tooltip,
                {
                  opacity: tooltipOpacity,
                  left: this.getTooltipXPosition(),
                  top: this.getTooltipYPosition(),
                },
              ]}
              onLayout={this.onTooltipLayout}
            >
              <Text style={{ color: Colors.white }}>{title}</Text>
            </View>
          )}
        </Portal>
        <View
          onTouchStart={this.onTouchStart}
          onTouchEndCapture={this.onTouchEndCapture}
          onTouchCancel={this.onTouchCancel}
        >
          {React.cloneElement(childElement, {
            onLayout: this.onChildrenLayout,
            onLongPress: () => {}, // Prevent touchable to trigger onPress after onLongPress
            ...rest,
          })}
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  tooltip: {
    alignSelf: 'flex-start',
    backgroundColor: Colors.grey700,
    borderRadius: 4,
    marginTop: 24,
    maxWidth: 300,
    paddingVertical: 4,
    paddingHorizontal: 8,
  },
});

export default Tooltip;
