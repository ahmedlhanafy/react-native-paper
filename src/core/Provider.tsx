import * as React from 'react';
import { Appearance, ColorSchemeName } from 'react-native';
import { ThemeProvider } from './theming';
import DefaultTheme from '../styles/DefaultTheme';
import DarkTheme from '../styles/DarkTheme';
import { Provider as SettingsProvider, Settings } from './settings';
import MaterialCommunityIcon from '../components/MaterialCommunityIcon';
import PortalHost from '../components/Portal/PortalHost';
import { Theme } from '../types';

type Props = {
  children: React.ReactNode;
  theme?: Theme;
  settings?: Settings;
};

type State = {
  colorScheme: ColorSchemeName;
};

export default class Provider extends React.Component<Props, State> {
  state = {
    colorScheme: Appearance.getColorScheme(),
  };

  componentDidMount() {
    Appearance.addChangeListener(this._handleAppearanceChange);
  }

  componentWillUnmount() {
    Appearance.removeChangeListener(this._handleAppearanceChange);
  }

  _handleAppearanceChange = (preferences: Appearance.AppearancePreferences) => {
    const { colorScheme } = preferences;
    this.setState({ colorScheme });
  };

  _getTheme = () => {
    if (this.props.theme) {
      return this.props.theme;
    } else {
      return this.state.colorScheme === 'light' ? DefaultTheme : DarkTheme;
    }
  };

  render() {
    return (
      <PortalHost>
        <SettingsProvider
          value={this.props.settings || { icon: MaterialCommunityIcon }}
        >
          <ThemeProvider theme={this._getTheme()}>
            {this.props.children}
          </ThemeProvider>
        </SettingsProvider>
      </PortalHost>
    );
  }
}
