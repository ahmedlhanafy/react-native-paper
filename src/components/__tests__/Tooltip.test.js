import React from 'react';
import { render, fireEvent } from 'react-native-testing-library';
import { View } from 'react-native';
import Tooltip from '../Tooltip';
import Provider from '../../core/Provider';

jest.mock('react-native/Libraries/ReactNative/UIManager', () => ({
  measure: () => {},
}));

jest.useFakeTimers();

describe('Tooltip', () => {
  it('renders properly', () => {
    const tree = render(
      <Tooltip title="Tooltip View">
        <View />
      </Tooltip>
    );

    expect(tree).toMatchSnapshot();
  });

  it('shows a tooltip when child element is longPressed', () => {
    const { getByTestId, toJSON } = render(
      <Provider>
        <Tooltip title="Tooltip View">
          <View testID="TooltipChild" />
        </Tooltip>
      </Provider>
    );

    fireEvent(getByTestId('TooltipChild'), 'touchStart');
    jest.runAllTimers();

    expect(toJSON()).toMatchSnapshot();
  });
});
