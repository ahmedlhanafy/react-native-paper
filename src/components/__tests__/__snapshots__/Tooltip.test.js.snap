// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders properly 1`] = `
<View>
  <View
    onTouchCancel={[Function]}
    onTouchEndCapture={[Function]}
    onTouchStart={[Function]}
  >
    <View
      delayLongPress={500}
      onLongPress={[Function]}
    />
  </View>
</View>
`;

exports[`Tooltip shows a tooltip when child element is longPressed 1`] = `
Array [
  <View
    collapsable={false}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        onTouchCancel={[Function]}
        onTouchEndCapture={[Function]}
        onTouchStart={[Function]}
      >
        <View
          delayLongPress={500}
          onLongPress={[Function]}
          testID="TooltipChild"
        />
      </View>
    </View>
  </View>,
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "alignSelf": "flex-start",
            "backgroundColor": "#616161",
            "borderRadius": 4,
            "marginTop": 24,
            "maxWidth": 300,
            "paddingHorizontal": 8,
            "paddingVertical": 4,
          },
          Object {
            "left": 0,
            "opacity": 0,
            "top": 0,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "Helvetica Neue",
              "textAlign": "left",
              "writingDirection": "ltr",
            },
            Object {
              "color": "#ffffff",
            },
          ]
        }
      >
        Tooltip View
      </Text>
    </View>
  </View>,
]
`;
